jupyterhub_chameleon/trovi.py

```diff
@@ -0,0 +1,88 @@
+import logging
+import os
+from urllib.parse import urljoin
+
+import requests
+from tornado.web import HTTPError
+
+LOG = logging.getLogger(__name__)
+
+def exchange_token(access_token):
+    # Exchange the user's keystone token for a Trovi token
+    token_resp = requests.post(
+        urljoin(os.getenv("TROVI_URL"), "/token/"),
+        headers={
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        },
+        json={
+            "grant_type": "token_exchange",
+            "subject_token": access_token,
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+            "scope": "artifacts:read artifacts:write",
+        },
+    )
+    LOG.debug(f"Trovi token exchange response: {token_resp}")
+
+    if token_resp.status_code in (
+            requests.codes.unauthorized,
+            requests.codes.forbidden,
+    ):
+        LOG.error(f"Authentication to trovi failed: {token_resp}")
+        raise HTTPError(
+            requests.codes.unauthorized,
+            "You are not authorized to upload artifacts to Trovi via Jupyter.",
+        )
+    elif token_resp.status_code != requests.codes.created:
+        LOG.error(f"Authentication to trovi failed: {token_resp}")
+        raise HTTPError(
+            requests.codes.internal_server_error,
+            requests.codes.internal_server_error,
+            "Unknown error uploading artifact to Trovi.",
+        )
+
+    return token_resp.json()
+
+def fetch_artifact(artifact_id, trovi_token):
+    artifact_req = requests.PreparedRequest()
```

Comment:
What is the reason for using PreparedRequest here? This is usually used for stuff like middleware, or when really specific/weird things need to be done to the URL. Neither seems to apply here, it looks like a pretty straightforward request to me.